### PR TITLE
Bare minimum z3 interface

### PR DESCRIPTION
**Pull request by tias**
Not implemented yet:
    - constraints : )
    - solving with assumptions
    - optimisation/objective

### CODE REVIEW

File: cpmpy/solvers/z3.py [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+        # we store them directly in `solver_var()` itself.
+        #self.user_vars.update(get_variables(cpm_con))
+
+        # only complication is that a list is implicitly an 'And' for us,
+        # as well as being a list of arguments, so top-level lists first
+        if is_any_list(cpm_con):
+            # recursively
+            for con in cpm_con:
+                self.__add__(con)
+        else:
```
Comment (IgnaceBleukx) [inline cpmpy/solvers/z3.py:228]: This isn't really compatible with our ideas for putting __add__ in the superclass and supporting by supporting a `transform()` function. 
I think the logic for posting the constraint to the z3 solver should still be contained in the `_post_constraint` method as is done in other solvers.

**Comment by IgnaceBleukx** [comment]
To implement optimization: we need the `Optimize` object instead of the `Solver` object of Z3. Not sure what the downsides are for using `Optimize` for SAT checking

**Comment by IgnaceBleukx** [comment]
Z3 doesn't directly support min/max consrtaints but it is quite easy to implement using upper/lower bounds on the arguments.
E.g.: `max(x,y,z) = m <=>  m >= x.ub & m >= y.ub & m >= z.ub`
So one reason more to fix #96 :)

**Comment by IgnaceBleukx** [comment]
> To implement optimization: we need the `Optimize` object instead of the `Solver` object of Z3. Not sure what the downsides are for using `Optimize` for SAT checking

to answer my own doubts:
https://stackoverflow.com/a/71754240/18784299

File: cpmpy/solvers/utils.py [hunk 2]
```diff
@@ -120,13 +123,12 @@ def lookup(name=None):
             solvername,_ = solvername.split(':',maxsplit=1)
 
         # find CPM_slv
```
Comment (tias) [inline cpmpy/solvers/utils.py:125]: everything below this is a bugfix, should be in a separate commit

File: cpmpy/solvers/z3.py [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+            sol = self.z3_solver.model() # the solution (called model in z3)
+            # fill in variable values
+            for cpm_var in self.user_vars:
+                sol_var = self.solver_var(cpm_var)
+                if isinstance(cpm_var, _BoolVarImpl):
+                    cpm_var._value = bool(sol[sol_var])
+                elif isinstance(cpm_var, _NumVarImpl):
+                    cpm_var._value = sol[sol_var].as_long()
+
+            # TODO
```
Comment (tias) [inline cpmpy/solvers/z3.py:163]: the TODO seems resolved?

File: cpmpy/solvers/z3.py (hunk 1 above)
Comment (tias) [reply cpmpy/solvers/z3.py:253]: I see, I agree. So yes, no transforms here and so this content would become post_constraint, go ahead

File: examples/advanced/VRP by learning from historical data.ipynb [hunk 4]
```diff
@@ -526,7 +526,7 @@
    "name": "python",
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.7.4"
+   "version": "3.9.2"
```
Comment (tias) [inline examples/advanced/VRP by learning from historical data.ipynb:529]: remove from commit

**Comment by tias** [comment]
Looks great, I left a few minor comments in the code (incl answer to your post_constraint question).

It looks feature complete, is it? If so, do merge after fixing the comments!

**Comment by IgnaceBleukx** [comment]
The reification of decomposable constraints is not fully working yet (such as min, max, abs). But this requires some more refactoring of the code I believe.

**Comment by tias** [comment]
I would be fine if you would throw 'NotImplementedErrors' for those cases and open a bugreport for it. We do need need feature completeness for it to be useful (e.g. no wsum in pysat yet, but still useful as is).
Then it can already be merged and used and tested, e.g. in our incremental work

File: cpmpy/solvers/z3.py (hunk 1 above)
Comment (IgnaceBleukx) [reply cpmpy/solvers/z3.py:261]: For future reference: the transform() function of z3 should then just make sure it returns a list of constraints
